pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -157,22 +161,77 @@ type DNSProxy struct {
 	unbindAddress func()
 }
 
+// patternMatcherEntry is a lookup entry used to cache a compiled regex
+// and how many references it has
+type patternMatcherEntry struct {
+	matcher        *regexp.Regexp
+	referenceCount int
+}
+
+// matcherCache is a reference counted cache used for reusing duplicate rules. This is useful when eg. multiple
+// pods have the same policy
+type matcherCache struct {
+	// mapping from regexp pattern to an actual regexp used for matchPatterns
+	patternMatcherCache map[string]*patternMatcherEntry
+}
+
 // perEPAllow maps EndpointIDs to ports + selectors + rules
 type perEPAllow map[uint64]portToSelectorAllow
 
 // portToSelectorAllow maps port numbers to selectors + rules
-type portToSelectorAllow map[uint16]CachedSelectorREEntry
+type portToSelectorAllow map[uint16]CachedSelectorEntry
+
+// fqdnMatcher is used to match an FQDN to see if the ruleset allows it or not.
+// In case a matcher is nil, it means that the respective matcher doesn't match any FQDNs
+type fqdnMatcher struct {
+	// patternMatcher is used to match an FQDN to the matchPatterns from a DNS policy
+	patternMatcher *regexp.Regexp
+}
+
+// CachedSelectorEntry maps port numbers to selectors to rules, mirroring
+// policy.L7DataMap but the DNS rules are compiled into a regex
+type CachedSelectorEntry map[policy.CachedSelector]fqdnMatcher
```

Comment:
We can flattern this right now to be `*regexp.Regexp` instead of `fqdnMatcher`. If/when we want the map/list of matchNames separately, we can introduce the `fqdnMatcher`

Reply:
Sure, let's flatten it now.